source/agora/network/CrawlerTypes.d

```diff
@@ -0,0 +1,98 @@
+/*******************************************************************************
+    Contains crawling related data types.
+
+    These types are used in the FullNode API, hence are separated from types
+    in Crawler.d to reduce compilation dependencies.
+
+    Copyright:
+        Copyright (c) 2019-2021 BOSAGORA Foundation
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.network.CrawlerTypes;
+
+import agora.common.Types : Address, TimePoint;
+import agora.serialization.Serializer;
+
+import std.algorithm : each;
+import std.range : iota;
+import std.traits : KeyType, ValueType;
+
+/// Result obtained by crawling a particular node
+public static struct CrawlResult
+{
+    ///
+    public string continent;
+
+    ///
+    public string country;
+
+    ///
+    public string city;
+
+    /// Latitude coordinates of the node
+    public string latitude;
+
+    /// Longitude coordinates of the node
+    public string longitude;
+
+    /// Operating system the node is running on
+    public string os;
+
+    /// The time the crawling result
+    public TimePoint crawl_time;
+}
+
+/// Crawling results for all the nodes that are directly or indirectly reachable
+/// from this node. Special serialize/fromBinary methods are provided, because
+/// the current binary serializer cannot deal with associative arrays.
+public static struct CrawlResultHolder
+{
+    /// Results for all the nodes that are directly or indirectly reachable
+    /// from this node
+    public CrawlResult[Address] crawl_results;
+
+    ///
+    alias crawl_results this;
+
+    ///
+    public void serialize (scope SerializeDg dg) const @safe
+    {
+        serializePart(cast(ulong) crawl_results.length, dg);
```

Comment:
not sure I understand this one... We need to know the size of the map to know how many elements we can read from it during deserialization.

Reply:
It's about the `cast`. `serializePart(crawl_results.length, dg);` should work just fine. The only reason I can think of for this `cast` is if we send that data on the network (and hence the other side might have `size_t.sizeof == 4`) or if we hash it.
In case we need it, using `serializePart!ulong` is the way to go as it relies on implicit conversion, and thus is more robust to change.

Reply:
length is size_t which is platform dependent that is why I used the cast

Reply:
Yes, so:
1) The data is sent on the network and using platform-dependent type is problematic (my original comment asked if that was the case);
2) You can just use explicit template instantiation to avoid the `cast` and make the code more robust.